core/src/doc/allow.rs

```diff
@@ -18,6 +18,20 @@ impl<'a> Document<'a> {
 		if self.id.is_some() {
 			// Should we run permissions checks?
 			if opt.check_perms(stm.into())? {
+				// Check that record authentication matches session
+				if opt.auth.is_record() {
+					let (ns, db) = (opt.ns()?, opt.db()?);
+					if opt.auth.level().ns() != Some(ns) {
+						return Err(Error::NsNotAllowed {
+							ns: ns.into(),
+						});
+					}
+					if opt.auth.level().db() != Some(db) {
+						return Err(Error::DbNotAllowed {
+							db: db.into(),
+						});
+					}
```

Comment:
if ns does not match, we still have retrieved opt.db.
As performance is a concern here,  could we do:

```suggestion
					if opt.auth.level().ns() != Some(opt.ns()?) {
						return Err(Error::NsNotAllowed {
							ns: ns.into(),
						});
					}
					if opt.auth.level().db() != Some(opt.db()?) {
						return Err(Error::DbNotAllowed {
							db: db.into(),
						});
					}
```

Reply:
Nice catch! You are absolutely right. Thanks a lot @emmanuel-keller. With your suggestion, the issue is that then `ns` and `db` do not have a value inside the `if` clause, which is why I assigned them to variables at the start. I don't know if there is a more Rust-ish way to do this, but I can just assign each variable separately before each condition.

Reply:
Fixed in [6653172](https://github.com/surrealdb/surrealdb/pull/4340/commits/6653172496fac305b331f3b67293382ab06e5a9c).